core/Enclave/key_factory.cpp

```diff
@@ -0,0 +1,570 @@
+
+    if (ret != SGX_SUCCESS)
+    {
+        goto out;
+    }
+out:
+    EVP_PKEY_CTX_free(pkey_ctx);
+    EVP_PKEY_free(pkey);
+    BIO_free(bio);
+    SAFE_FREE(pem_keypair);
+    return ret;
+}
+
+sgx_status_t ehsm_create_ec_key(ehsm_keyblob_t *cmk)
+{
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
+
+    if (cmk == NULL)
+    {
+        return ret;
+    }
+
+    if (cmk->keybloblen == 0)
+    {
+        cmk->keybloblen = PEM_BUFSIZE;
+        return SGX_SUCCESS;
+    }
+
+    EVP_PKEY_CTX *pkey_ctx = NULL;
+    EVP_PKEY *pkey = NULL;
+    BIO *bio = NULL;
+    uint8_t *pem_keypair = NULL;
+    uint32_t key_len = 0;
+
+    pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_EC, NULL);
+    if (pkey_ctx == NULL)
+    {
+        goto out;
+    }
+
+    if (EVP_PKEY_keygen_init(pkey_ctx) <= 0)
+    {
+        goto out;
+    }
+
+    switch (cmk->metadata.keyspec)
+    {
+    case EH_EC_P224:
+    case EH_EC_P256:
+    case EH_EC_P384:
+    case EH_EC_P512:
+        if (EVP_PKEY_CTX_set_ec_paramgen_curve_nid(pkey_ctx, NID_X9_62_prime256v1) <= 0)
+        {
+            goto out;
+        }
+        break;
+    case EH_SM2:
+        if (EVP_PKEY_CTX_set_ec_paramgen_curve_nid(pkey_ctx, NID_sm2) <= 0)
+        {
+            goto out;
+        }
+        break;
+    default:
+        break;
+    }
+
+    if (EVP_PKEY_keygen(pkey_ctx, &pkey) <= 0)
+    {
+        goto out;
+    }
+
+    bio = BIO_new(BIO_s_mem());
+    if (bio == NULL)
+    {
+        goto out;
+    }
+
+    if (!PEM_write_bio_PUBKEY(bio, pkey))
+    {
+        goto out;
+    }
+
+    if (!PEM_write_bio_PrivateKey(bio, pkey, NULL, NULL, 0, NULL, NULL))
+    {
+        goto out;
+    }
+
+    key_len = BIO_pending(bio);
+    if (key_len <= 0)
+    {
+        goto out;
+    }
+
+    pem_keypair = (uint8_t *)malloc(key_len + 1);
+    if (pem_keypair == NULL)
+    {
+        goto out;
+    }
+
+    if (BIO_read(bio, pem_keypair, key_len) < 0)
+    {
+        goto out;
+    }
+    pem_keypair[key_len] = '\0';
+
+    ret = ehsm_create_keyblob(pem_keypair, key_len, NULL, 0, (sgx_aes_gcm_data_ex_t *)cmk->keyblob);
+
+    if (ret != SGX_SUCCESS)
+    {
+        goto out;
+    }
+out:
+    EVP_PKEY_CTX_free(pkey_ctx);
+    EVP_PKEY_free(pkey);
+    BIO_free(bio);
+    SAFE_FREE(pem_keypair);
+    return ret;
+}
+
+sgx_status_t ehsm_create_sm4_key(ehsm_keyblob_t *cmk)
+{
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
+
+    if (cmk->metadata.keyspec != EH_SM4_CTR &&
+        cmk->metadata.keyspec != EH_SM4_CBC)
+    {
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
+
+    uint32_t keysize = ehsm_get_symmetric_key_size(cmk->metadata.keyspec);
+    if (keysize == 0)
+    {
+        return SGX_ERROR_UNEXPECTED;
+    }
+    uint32_t real_blob_len = ehsm_calc_keyblob_len(cmk->metadata.keyspec);
+
+    if (real_blob_len == UINT32_MAX)
+    {
+        return SGX_ERROR_UNEXPECTED;
+    }
+    if (cmk->keybloblen == 0)
+    {
+        cmk->keybloblen = real_blob_len;
+        return SGX_SUCCESS;
+    }
```

Comment:
please follow the same code logic for other function to calculate the length and return to caller

Reply:
copy

Reply:
please double check each create_xxx interface and keep the same coding style, currently it's not same